### PR TITLE
Rewrite and Refractor

### PR DESCRIPTION
**Pull request by ItayZiv**
Part of https://github.com/wpilibsuite/sphinxext-opengraph/issues/11.
TODOs can be ignored (will be removed).

### CODE REVIEW

File: dev-requirements.txt [hunk 1]
```diff
@@ -1,5 +1,6 @@
-sphinx
+sphinx==3.1.2
```
Comment (Daltz333) [inline dev-requirements.txt:1]: We should always be on the latest version, we want to know when Sphinx does a breaking change.

File: sphinxext/opengraph/util.py [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+        desc_len = int(config["ogp_description_length"])
+    except ValueError:
+        desc_len = 300
+
+    # parse out any html from the title
+    html_parser = HTMLTextParser()
+    if context["title"] != "&lt;no title&gt;":
+        # only use title if its exists
+        html_parser.feed(context["title"])
+    html_parser.close()
```
Comment (TheTripleV) [inline sphinxext/opengraph/util.py:27]: I know this is this original implementation. But, I think it would be better as a function.

```py
def santitize_title(title: str) -> str:
	# take in a title and return a clean version.
```

Also, `<no title>` shouldn't be an edge case. That is the page title as used by Sphinx.

File: sphinxext/opengraph/visitor.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+            pass
+
+        if isinstance(node, (nodes.Invisible, nodes.Admonition)):
+            # Skip all comments and admonitions
+            raise nodes.SkipNode
+
+        if isinstance(node, nodes.title):
+            # title count refers to the amount of titles remaining
+            self.title_count -= 1
+            if self.title_count < 0:
```
Comment (TheTripleV) [inline sphinxext/opengraph/visitor.py:27]: `count`s should count upwards.

Also, what if the first section does not have text?
Example: https://docs.wpilib.org/en/latest/docs/software/driverstation/manually-setting-the-driver-station-to-start-custom-dashboard.html

If we start counting sections, we should either:
1. Start counting at the first section containing text
2. Only count sections containing text.

File: sphinxext/opengraph/visitor.py [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    def __init__(
+        self, document: nodes.document, desc_len: int, title_count: int = 1
+    ) -> None:
+        super().__init__(document)
+        self.description = ""
+        self.desc_len = desc_len
+        self.title_count = title_count
+        self.first_list = ""  # todo: rename
+
+    def default_visit(self, node: nodes.Element) -> None:
```
Comment (TheTripleV) [inline sphinxext/opengraph/visitor.py:14]: For Discord question 3, I think if statements are better here than `'visit_' + node_name` functions. So, as is currently.

File: sphinxext/opengraph/event_handler.py [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+
+def html_page_context(
+    app: Sphinx,
+    pagename: str,
+    templatename: str,
+    context: Dict[str, Any],
+    doctree: nodes.document,
+) -> None:
+    if doctree:
+        insert_tags(context, doctree, app.config)
```
Comment (TheTripleV) [inline sphinxext/opengraph/event_handler.py:15]: `html_page_context` can be combined with `insert_tags` since it (`html_page_context`) is just one if statement.

File: sphinxext/opengraph/util.py (hunk 2 above)
Comment (ItayZiv) [reply sphinxext/opengraph/util.py:27]: I already removed `<no title>`, haven't pushed yet.
And that function would go in `util.py` right?

File: sphinxext/opengraph/visitor.py (hunk 3 above)
Comment (ItayZiv) [reply sphinxext/opengraph/visitor.py:27]: I forgot to ask about this part in the discord, but this is another part I wasn't really sure about.
Right now the idea is the user can enter how many sections to search, but another possibility would be to only grab the first section that isn't empty.
I agree that regardless empty sections should be ignored.
Not sure if we should put in the titles of empty sections or ignore all titles as it is now.

File: sphinxext/opengraph/event_handler.py (hunk 5 above)
Comment (ItayZiv) [reply sphinxext/opengraph/event_handler.py:15]: Probably, this was that way is was originally so I just kinda left it that way, will change

File: dev-requirements.txt (hunk 1 above)
Comment (ItayZiv) [reply dev-requirements.txt:1]: We could have dependabot bump the version when new ones come out or something since this would mean that at some point there could be a PR whos tests won't pass because tests on master don't pass... in general, a test's result should always be reproducible IMO.

File: sphinxext/opengraph/__init__.py [hunk 6]
```diff
@@ -0,0 +1,23 @@
+from typing import Any, Dict
+from sphinx.application import Sphinx
+from .event_handler import html_page_context
+
+# todo: rtd option to grab READTHEDOCS_LANGUAGE and READTHEDOCS_VERSION for link
```
Comment (Daltz333) [inline sphinxext/opengraph/__init__.py:5]: We can put these TODOs in a separate PR after this is merged.

File: sphinxext/opengraph/util.py (hunk 2 above)
Comment (TheTripleV) [reply sphinxext/opengraph/util.py]: Yeah, util would be a good place for it.

File: sphinxext/opengraph/__init__.py (hunk 6 above)
Comment (ItayZiv) [reply sphinxext/opengraph/__init__.py:5]: Yea, those are going to be gone before I remove it from being a draft, just waiting for an opinion on the lists and sections to close that part off

File: sphinxext/opengraph/visitor.py (hunk 3 above)
Comment (TheTripleV) [reply sphinxext/opengraph/visitor.py:27]: Regarding titles, the options I see are either:
1. Only include titles of sections with text
2. Ignore all titles.

Either is fine really.

Regarding counting sections:
Counting sections falls counter to description length. If the user specifies 300 characters, then it may not be intuitive as to why a lot less characters were included.

I'm fine with providing options for both description_length and num_sections but it would need to be clear in the readme that the amount of text grabbed is `pseudo_min(description_length, num_sections)`.

**Comment by TheTripleV** [comment]
Regarding list handling:
All I care about is that index pages (that only have lists) have lists formatted.

The one edge case I can think of that needs to be clarified:
https://docs.wpilib.org/en/latest/docs/getting-started/getting-started-frc-control-system/radio-programming.html

It's a list in the first section of text.

File: sphinxext/opengraph/visitor.py (hunk 3 above)
Comment (ItayZiv) [reply sphinxext/opengraph/visitor.py:27]: Regarding the counting sections, my train of thought is that `ogp:description` is meant to be a small short description of the page, just grabbing the first X chars means it will always just be everything that can fit, while grabbing the first paragraph will most likely be some kind of either description of the page, or snippet of the contents of the page.
I wouldn't be opposed to having it only grab the first section with text and renaming the `desc_len` to `max_desc_len`.
The other option would be to have that configurable, where it can grab the first 300 chars and ignore all titles, or do as mentioned above.
We could get into more complex title impls, but I feel like the thing I said above should be the default and be the best in most use cases.

**Comment by ItayZiv** [comment]
My thought is to go back to a more similar impl to what was before, just actually format lists in. This is, numbered lists are treated like a normal list, sublists have a bit of special formatting, the rest is just comma-separated elements.
It *might* be worth ignoring sections with only lists, but not sure if that's something that should be dealt with...

File: sphinxext/opengraph/visitor.py (hunk 3 above)
Comment (TheTripleV) [reply sphinxext/opengraph/visitor.py:27]: Grabbing just one section isn't enough. The previous impl by a Sphinx Maintainer was along the lines of `pseudomin(3 sections, 200 char)`. I agree with your argument for counting sections. I just think that both should be options.

**Comment by TheTripleV** [comment]
I don't really have a strong opinion on lists other than the point about index pages I made above. As long as that works, do what you think is best.

File: sphinxext/opengraph/event_handler.py [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+    )
+
+    # add all the tags which need to be added
+    context["metatags"] += make_tag("title", page_title)
+    context["metatags"] += make_tag("type", app.config["ogp_type"])
+    context["metatags"] += make_tag("url", page_url)
+    add_tag_from_config(context, app.config, "site_name")
+    context["metatags"] += make_tag("description", visitor.description)
+    add_tag_from_config(context, app.config, "image")
+    add_tag_from_config(context, app.config, "image:alt")
```
Comment (TheTripleV) [inline sphinxext/opengraph/event_handler.py:43]: I don't like the mixed usage of `context["metatags"] += ` and `add_tag_from_config`. I think `add_tag_from_config` reduces readability and should be removed.

File: sphinxext/opengraph/event_handler.py [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+    context: Dict[str, Any],
+    doctree: nodes.document,
+) -> None:
+    if not doctree:
+        return
+    # Set length of description
+    try:
+        desc_len = int(app.config["ogp_description_length"])
+    except ValueError:
+        desc_len = 200
```
Comment (TheTripleV) [inline sphinxext/opengraph/event_handler.py:22]: Constants shouldn't be defined twice. 
Since `app.add_config_value("ogp_description_length", 200, "html")` is already in `__init__.py`, this isn't necessary.
